src/out_ref.rs

```diff
@@ -0,0 +1,573 @@
+    /// );
+    /// // we can thus soundly `assume_init` our array:
+    /// let array = unsafe {
+    ///     mem::transmute::<
+    ///         [MaybeUninit<u8>; 13],
+    ///         [            u8 ; 13],
+    ///     >(array)
+    /// };
+    /// assert_eq!(
+    ///     array,
+    ///     *b"Hello, World!",
+    /// );
+    /// ```
+    pub
+    fn copy_from_slice (
+        mut self: Out<'out, [T]>,
+        source_slice: &'_ [T],
+    ) -> &'out mut [T]
+    where
+        T : Copy,
+    {
+        unsafe {
+            // # Safety
+            //
+            //   - Writing to `self.0` is fine since `source_slice` only
+            //     contains initialized elements.
+            //
+            //   - the `copy_nonoverlapping()` call guarantees that the buffer
+            //     has been initialized.
+            self.r()
+                .as_uninit_mut()
+                .copy_from_slice(
+                    <[MaybeUninit<T>]>::from_ref(source_slice)
+                )
+            ;
+            self.assume_init()
+        }
+    }
+
+    /// Fill the buffer with values from a function mapping indices to values.
+    ///
+    /// # Guarantees
+    ///
+    /// `factory` is guaranteed to be called in the `0 .. self.len()` order.
+    #[inline]
+    pub
+    fn fill_with (
+        mut self: Out<'out, [T]>,
+        mut factory: impl FnMut(usize) -> T,
+    ) -> &'out mut [T]
+    {
+        self.iter_out()
+            .enumerate()
+            .for_each(|(i, at_dst)| { at_dst.write(factory(i)); })
+        ;
+        unsafe {
+            // Safety: The `len` values of the buffer have been initialized
+            self.assume_init()
+        }
+    }
+
+    /// `.reborrow().into_iter()`
+    #[inline]
+    pub
+    fn iter_out<'reborrow> (self: &'reborrow mut Out<'out, [T]>)
+      -> iter::IterOut<'reborrow, T>
+    {
+        self.r().into_iter()
+    }
+
+    #[inline]
+    pub
+    fn split_at_out (self: Out<'out, [T]>, idx: usize)
+      -> (Out<'out, [T]>, Out<'out, [T]> )
+    {
+        let (left, right) = unsafe { self.as_uninit_mut() }.split_at_mut(idx);
+        (left.into(), right.into())
+    }
+}
+
+/// `Deref` into `[MaybeUninit<T>]` to get access to the slice length related
+/// getters.
+impl<'out, T : 'out> ::core::ops::Deref for Out<'out, [T]> {
+    type Target = [MaybeUninit<T>];
+
+    #[inline]
+    fn deref (self: &'_ Self)
+      -> &'_ [MaybeUninit<T>]
+    {
+        unsafe {
+            &*(self.0.as_ptr() as *const [MaybeUninit<T>])
+        }
+    }
+}
+
+use private::SliceIndex;
+mod private {
+    use super::*;
+
+    pub
+    trait SliceIndex<'out, T> {
+        type Output : 'out;
+
+        fn idx (self: Self, slice: Out<'out, [T]>)
+          -> Option<Self::Output>
+        ;
+    }
+}
+
+pub
+mod iter {
+    use super::*;
+
+    pub
+    struct IterOut<'out, T : 'out> {
+        /// Points to the remaining elements of the iterator.
+        pub
+        slice: Out<'out, [T]>,
```

Comment:
A public member in an iterator? That's quite unorthodox especially since the slice can be reborrowed. How about adding a `get_out` method/impl instead?

Reply:
Note: being able to set the member could have some implications with regards to `ExactSizeIterator` and other iterator traits or specialization implementations. It's not 100% clear to me though.